### PR TITLE
US016 edit capability

### PR DESCRIPTION
**Pull request by caitlinbulloch**
**Ability to Edit a Capability Name**

- Created capability Dao.
- Created capability Service.
- Created capability controller.
- Added appropriate exceptions.
- Created unit tests on 'updateCapability'.

Screenshot 2023-12-21 at 10 06 58 (2)

### CODE REVIEW

**Comment by CharlesM988** [comment]
Run and looks good for me. I can't see anything that stands out.

Screenshot 2024-01-09 at 09 44 45
Screenshot 2024-01-09 at 09 45 06

File: src/main/java/org/kainos/ea/api/CapabilityService.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            throw new FailedToGetCapabilitiesException();
+        }
+
+        return capabilityList;
+
+    }
+    public Capability getCapabilityByID(int id) throws FailedToGetCapabilityException {
+        Capability capability;
+        try {
+            capability = capabilityDao.getCapabilityByID(id);
```
Comment (eoinmcbennett) [inline src/main/java/org/kainos/ea/api/CapabilityService.java]: Could we not just return here the value from getCapabilityByID directly?

`return getCapabilityByID(id)`

File: src/main/java/org/kainos/ea/api/CapabilityService.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
+
+            throw new FailedToGetCapabilityException();
+        }
+        return capability;
+    }
+
+    public void updateCapability(int id, CapabilityRequest capability) throws CapabilityDoesNotExistException,
+            FailedToUpdateCapabilityException, InvalidCapabilityException, SQLException {
```
Comment (eoinmcbennett) [inline src/main/java/org/kainos/ea/api/CapabilityService.java]: Does it make sense for this method to throw a SQL exception? Could we handle it in here and throw something more descriptive back to the caller?

File: src/main/java/org/kainos/ea/client/CapabilityValidator.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package org.kainos.ea.client;
+
+import org.kainos.ea.cli.CapabilityRequest;
+
+public class CapabilityValidator {
+    private final int maxLength = 100;
```
Comment (eoinmcbennett) [inline src/main/java/org/kainos/ea/client/CapabilityValidator.java]: Constants in java should be all caps. e.g MAX_LENGTH

File: src/main/java/org/kainos/ea/resources/CapabilityController.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        try {
+            capabilityService.updateCapability(id, capability);
+
+            return Response.ok().build();
+        } catch (CapabilityDoesNotExistException | InvalidCapabilityException e) {
+            System.err.println(e.getMessage());
+
+            return Response.status(Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
+
+        } catch (FailedToUpdateCapabilityException | SQLException e) {
```
Comment (eoinmcbennett) [inline src/main/java/org/kainos/ea/resources/CapabilityController.java]: Can a SQL exception ever get to this point?

File: src/main/java/org/kainos/ea/resources/CapabilityController.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response updateCapability(@PathParam("id") int id, CapabilityRequest capability) {
+        try {
+            capabilityService.updateCapability(id, capability);
+
+            return Response.ok().build();
+        } catch (CapabilityDoesNotExistException | InvalidCapabilityException e) {
+            System.err.println(e.getMessage());
+
+            return Response.status(Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
```
Comment (eoinmcbennett) [inline src/main/java/org/kainos/ea/resources/CapabilityController.java:67]: Can this message ever be something sensitive that we don't want the user seeing?

**Review by eoinmcbennett — CHANGES_REQUESTED** [review]

**Review by eoinmcbennett — APPROVED** [review]

**Review by CharlesM988 — APPROVED** [review]
LGTM

**Pull request merged**